lib/SIL/MemAccessUtils.cpp

```diff
@@ -281,174 +281,218 @@ class AccessedStorageResult {
-  AccessedStorage::Kind kind = AccessedStorage::classify(sourceAddr);
-  // First handle identified cases: these are always valid as the base of
-  // a formal access.
-  if (kind != AccessedStorage::Unidentified)
-    return AccessedStorage(sourceAddr, kind);
-
-  // If the sourceAddr producer cannot immediately be classified, follow the
-  // use-def chain of sourceAddr, box, or RawPointer producers.
-  assert(sourceAddr->getType().isAddress()
-         || isa<SILBoxType>(sourceAddr->getType().getASTType())
-         || isa<BuiltinRawPointerType>(sourceAddr->getType().getASTType()));
-
-  // Handle other unidentified address sources.
-  switch (sourceAddr->getKind()) {
-  default:
-    if (isAddressForLocalInitOnly(sourceAddr))
-      return AccessedStorage(sourceAddr, AccessedStorage::Unidentified);
-    return AccessedStorage();
-
-  case ValueKind::SILUndef:
-    return AccessedStorage(sourceAddr, AccessedStorage::Unidentified);
-
-  case ValueKind::ApplyInst:
-    if (isExternalGlobalAddressor(cast<ApplyInst>(sourceAddr)))
-      return AccessedStorage(sourceAddr, AccessedStorage::Unidentified);
-
-    // Don't currently allow any other calls to return an accessed address.
-    return AccessedStorage();
-
-  case ValueKind::StructExtractInst:
-    // Handle nested access to a KeyPath projection. The projection itself
-    // uses a Builtin. However, the returned UnsafeMutablePointer may be
-    // converted to an address and accessed via an inout argument.
-    if (isUnsafePointerExtraction(cast<StructExtractInst>(sourceAddr)))
-      return AccessedStorage(sourceAddr, AccessedStorage::Unidentified);
-    return AccessedStorage();
-
-  case ValueKind::SILPhiArgument: {
-    auto *phiArg = cast<SILPhiArgument>(sourceAddr);
-    if (phiArg->isPhiArgument())
-      return AccessedStorageResult::incomplete(phiArg);
-
-    // A non-phi block argument may be a box value projected out of
-    // switch_enum. Address-type block arguments are not allowed.
-    if (sourceAddr->getType().isAddress())
-      return AccessedStorage();
-
-    checkSwitchEnumBlockArg(cast<SILPhiArgument>(sourceAddr));
-    return AccessedStorage(sourceAddr, AccessedStorage::Unidentified);
-  }
-  // Load a box from an indirect payload of an opaque enum.
-  // We must have peeked past the project_box earlier in this loop.
-  // (the indirectness makes it a box, the load is for address-only).
-  //
-  // %payload_adr = unchecked_take_enum_data_addr %enum : $*Enum, #Enum.case
-  // %box = load [take] %payload_adr : $*{ var Enum }
-  //
-  // FIXME: this case should go away with opaque values.
-  //
-  // Otherwise return invalid AccessedStorage.
-  case ValueKind::LoadInst:
-    if (sourceAddr->getType().is<SILBoxType>()) {
-      SILValue operAddr = cast<LoadInst>(sourceAddr)->getOperand();
-      assert(isa<UncheckedTakeEnumDataAddrInst>(operAddr));
-      return AccessedStorageResult::incomplete(operAddr);
+class AccessUseDefChainVisitor {
```

Comment:
Maybe we could make the implementation of `visitAccessUseDefChain` `always_inline` within MemAccessUtils.cpp, so that the visitor can be devirtualized.